docs/Guide.TypeScriptIntegrationJest.md

```diff
@@ -0,0 +1,120 @@
+---
+id: Guide.TypeScriptJest
+title: Integrating TypeScript with Detox + Jest
+---
+
+## Usage
+
+### 0. Initialize Detox in your project
+
+After following the [Getting Started](Introduction.GettingStarted.md) Guide to set up Detox + Jest, we will make some changes to your project:
+
+- Remove Detox from `package.json` at the root level: `npm uninstall detox`
+- Create a `package.json` in the `e2e` folder and add Detox and Jest to the project dependencies:
+
+```sh
+cd e2e
+npm init -y
```

Comment:
I am very confused about this step, why would the users need to uninstall Detox from the main project and init a new project inside e2e dir? This looks like a hack :(
Is it only because of the detox globals?

Reply:
I think this is a hacky solution and can not be documented as an official solution. There must be a better way.

Reply:
Can you elaborate on why it's a hacky solution, and how it can be improved?

Reply:
Setting Detox as a devDependency of a subproject inside e2e folder, running npm install twice is not a very common, or even wanted practice, it can and will have grave consequences on projects maintenance, especially if they're big, or include many developers.
I'm not an expert in typescript configuration, but this solution, as it is a bit hacky, can not be part of official documentation.
We should strive to solve the real problem, not to document workarounds.

Now, how can we solve this?

Reply:
I don't seeing any consequences of having multiple `package.json`s since there shouldn't be overlap between e2e code and implementation code. As I pointed out above, I do have an example project that has only one `package.json` [here](https://github.com/solkaz/ts-detox-example/tree/one-package-json-to-rule-them-all) and mimics the usual Detox setup (I think). If you're okay with that setup, then I'll update the PR to reflect that.

Reply:
> Now, how can we solve this?

It seems like the 2nd package.json in e2e/ is only really necessary to avoid typing conflicts with `expect`, `describe` and other globals that Jest and Detox share. Knowing this, could a potential solution be that we are required to import the globals from their respective lib before using them? For example, `import { expect, describe, it, element } from "detox";`

I realize the scope of this change is large, and would have to be on an opt-in basis to avoid affecting existing Detox test suites. I am sure that I am missing some background as to why the global approach was taken, since I feel like most React Native devs would be comfortable with importing their dependencies.

> I do have an example project that has only one package.json here

@solkaz, this looks good to me! One less `npm install` to run, one less `node_modules` directory full of dependencies.

Reply:
Dug a bit more and discovered that this already exists! https://github.com/wix/Detox/blob/master/docs/APIRef.DetoxObjectAPI.md#detoxinit

In init.js|ts:
```await detox.init(config, {initGlobals: false});```

At the top of each .spec.ts:
```const {device, expect, element, by, waitFor} = require('detox');```

Cool!

Reply:
This was already [mentioned in the thread ](https://github.com/wix/Detox/pull/1001#discussion_r228734569) but was said to still suffer from typing issues.

Reply:
@blefebvre to expand on what @rotemmiz said, once [this PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30410) goes through, you would be able to import the constants from `detox`:

```ts
const { device, expect, element, by, waitFor } from "detox";
```

TypeScript will recognize `expect` to be a `Detox.Expect` so your files would then compile. Currently this isn't possible because Detox isn't declared as a module, and `require`-ing doesn't seem to work because `expect` has already been declared.

Reply:
@rotemmiz I've made changes to the document that recommend disabling `initGlobals` and using import statements for the Detox methods.